practice/2021/armine_vartanova/AbstractFactory.cpp

```diff
@@ -0,0 +1,235 @@
+#include <iostream>
+#include <map>
+using namespace std;
+
+/*
+Not necessary to have a base class for armor parts, but we're
+including it since all the armor parts have the same functionality
+ */
+class Armor {
+protected: int defence;
+public:
+    void TakeDamage(int damage)
+    {
+        defence -= damage;
+    }
+protected:
+    void setDefence();
+};
+
+//Product family
+class Gauntlets : public Armor
+{
+   void setDefence(){
+       defence = 10;
+    }
+};
+class Headgear : public Armor
+{
+    void setDefence(){
+        defence = 20;
+     }
+};
+class Cuirass : public Armor
+{
+    void setDefence(){
+        defence = 50;
+     }
+};
+class Boots : public Armor
+{
+    void setDefence(){
+        defence = 30;
+     }
+};
+
+//Product subclass family
+class DragonboneGauntlets : public Gauntlets {
+    void setDefence(){
+        defence = 20;
+     }
+};
+class DragonboneHeadgear : public Headgear {
+    void setDefence(){
+        defence = 30;
+     }
+};
+class DragonboneCuirass : public Cuirass {
+    void setDefence(){
+        defence = 60;
+     }
+};
+class DragonboneBoots : public Boots {
+    void setDefence(){
+        defence = 40;
+     }
+};
+
+//Product subclass family
+class ElvenGauntlets : public Gauntlets {
+    void setDefence(){
+        defence = 15;
+     }
+};
+class ElvenHeadgear : public Headgear {
+    void setDefence(){
+        defence = 25;
+     }
+};
+class ElvenCuirass : public Cuirass {
+    void setDefence(){
+        defence = 55;
+     }
+};
+class ElvenBoots : public Boots {
+    void setDefence(){
+        defence = 35;
+     }
+};
+
+enum ArmorPiece { HEADGEAR, BOOTS, CUIRASS, GAUNTLETS };
+
+class Equipment {
+private: map<ArmorPiece,Armor*> _equipment;
```

Comment:
Replace this map by simple Armor piece fields (one for each type, e.g. one Headgear).